src/utils/priority-queue.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A priority queue backed with sorted array.
+ * @template T
+ */
+export default class PriorityQueue {
+  constructor() {
+    /** @private @const {Array<{item: T, priority: number}>} */
+    this.queue_ = [];
```

Comment:
We'd get theoretically better performance with a max heap for large N but I thought it was unnecessary for current usage. I'm not sure that even the binary search is worth the readability cost.